### DISCUSSION TITLE
Dataloaders vs SelectFields

### DISCUSSION THREAD

**Discussion by asanovr** [rebing/graphql-laravel#946] [Q&A]
Hi! I've read about SelectFields in issues and discussions and some developers prefer dataloaders. 
What are any pros and cons of each approach?
Currently, I've got a problem with using ::select and ::with methods in the same query, and I'm looking for some solutions.

**Comment by mfn** [comment]
`SelectFields` inspect the GraphQL nested fields being sent and then deduce an Eloquent query from that; it also supports loading relation onto it (`with()`) and even supports custom queries on _them_.

Whilst this sounds perfect and works for many easy cases, at some point one will realize its shortcomings, that certain queries and the way the model system is setup will not work and not produce the desired result.

At this point, you've to options: fix `SelectFields` or use a dataloader approach.

The dataloader approach itself is just an architectural concept and this library does not dictate whether you use overblog/dataloader-php or roll your own. All you do with such concepts is, you "write your own resolvers" everywhere. Which can be a lot of bloat for easy tasks (especially relations) but OTOH gives you complete freedom and ways of optimizing your results you cannot do with `SelectFields`.

Further, `SelectFields` also mixes GraphQL type system fields with its own which can cause confusion when does what apply.

TL;DR:
- SelectFields: quick to get started, shortcomings probably never solvable
- dataloader: more work but as flexible as any code you write

Personally I never used SelectFields, always dataloader approach (we rolled our own implementation though). And I'm still using this library in production, so it's not just some distant memory either.

HTH!

**Reply by asanovr** [reply]
Thank you for the answer! 
Could you describe how you work with eloquent::select and dataloaders?
Do you have any examples of dataloaders? I think it will be very useful.

**Reply by mfn** [reply]
This library is just a wrapper around graphql-php, they've a section about how to apply it -> https://webonyx.github.io/graphql-php/data-fetching/#solving-n1-problem . Here is a very popular library, but as I said, I'm using my own thing -> https://github.com/overblog/dataloader-php

You might also want to check out Lighthouse, which is also built on top of graphql-php but has a different approach and probably feels "more integrated" with Eloquent too -> https://lighthouse-php.com/ . It's using a "schema first" approach, so it works a bit different.
